### DISCUSSION TITLE
Extending light/dark change detection to fish shell with `os.execute`

### DISCUSSION THREAD

**Discussion by simonmandlik** [wezterm/wezterm#5951] [Q&A]
### What Operating System(s) are you running on?

macOS

### WezTerm version

20240203-110809-5046fc22

### Ask your question!

Hi!

I'm trying to extend the OS appearance detection as demoed [here](https://wezfurlong.org/wezterm/config/lua/wezterm.gui/get_appearance.html#wayland-gnome-appearance)

to fish shell. In my fish config I have a function that gets called whenever universal variable `OS_APPEARANCE` changes value:

```fish
function update_theme --on-variable OS_APPEARANCE
    echo "$OS_APPEARANCE"
    # ... logic setting the theme based on OS_APPEARANCE
end
```

This can be set from anywhere (even different shell or terminal) with

`fish -c 'set --universal OS_APPEARANCE {light, dark}'`

I tested this in different terminals and shells and even in lua:

```bash
lua -e "os.execute([[fish -c 'set --universal OS_APPEARANCE dark']])"
```

and the call works every time.

However, when I put it inside `~/.wezterm.lua` like this:

```lua
local wezterm = require 'wezterm'

-- wezterm.gui is not available to the mux server, so take care to
-- do something reasonable when this config is evaluated by the mux
function get_appearance()
    if wezterm.gui then
        return wezterm.gui.get_appearance()
    end
    return 'Dark'
end

function scheme_for_appearance(appearance)
    if appearance:find 'Dark' then
        os.execute("fish -c 'set --universal OS_APPEARANCE dark'")
        return 'Builtin Solarized Dark'
    else
        os.execute("fish -c 'set --universal OS_APPEARANCE light'")
        return 'Builtin Solarized Light'
    end
end

return {
    color_scheme = scheme_for_appearance(get_appearance()),
}
```

fish never reacts. Is there a way to get around this?

Many thanks!

**Comment by wez** [comment]
First take a look at:
* https://wezfurlong.org/wezterm/faq.html#im-on-macos-and-wezterm-cannot-find-things-in-my-path

Second: that config will spawn a fish subprocess every time the config file is evaluated. 
https://wezfurlong.org/wezterm/config/files.html#configuration-files says:

> The configuration file may be evaluated multiple times for each wezterm process both at startup and in response to the configuration file being reloaded. You should avoid taking actions in the main flow of the config file that have side effects; for example, unconditionally launching background processes can result in many of them being spawned over time if you launch many copies of wezterm, or are frequently reloading your config file.

so I would strongly recommend against doing that.

Look at using https://wezfurlong.org/wezterm/config/lua/wezterm/GLOBAL.html to remember the last observed value for the appearance and only spawn fish when the value changes.

Thirdly: `os.execute` can block the gui thread if the command isn't guaranteed to always return quickly. For that reason, consider using https://wezfurlong.org/wezterm/config/lua/wezterm/background_child_process.html instead of `os.execute`.

**Reply by simonmandlik** [reply]
Thanks for the help!

I followed all the tips and ended up with this simple config

```lua
local wezterm = require 'wezterm'

local function get_appearance()
    if wezterm.gui then
        if wezterm.gui.get_appearance():find 'Dark' then
            return 'Dark'
        else
            return 'Light'
        end
    end
    return 'Dark'
end

local appearance = get_appearance()
if wezterm.GLOBAL.appearance ~= appearance then
    wezterm.GLOBAL.appearance = appearance
    local cmd = [[/usr/local/Cellar/fish/3.7.1/bin/fish -c "set --universal OS_APPEARANCE ]] .. appearance .. [["]]
    wezterm.background_child_process { cmd }
end

return {}
```

using the absolute path to fish shell with all symlinks resolved.

Unfortunately, I still get this error:

```
No such file or directory (os error 2)
stack traceback:
        [C]: in local 'poll'
        [string "?"]:4: in function 'wezterm.background_child_process'
        [string "/Users/simon.mandlik/.wezterm.lua"]:19: in main chunk
```

**Reply by wez** [reply]
You need to pass a command array with the first arg being the path to fish:

```lua
 local cmd = {'/usr/local/Cellar/fish/3.7.1/bin/fish', '-c', 'set --universal OS_APPEARANCE ' .. appearance }
```

**Reply by simonmandlik** [reply]
Amazing, thanks!

If anyone else struggles with this, this is a complete working config:

```lua
local wezterm = require 'wezterm'

local function get_appearance()
    if wezterm.gui then
        if wezterm.gui.get_appearance():find 'Dark' then
            return 'Dark'
        else
            return 'Light'
        end
    end
    return 'Dark'
end

local appearance = get_appearance()
if wezterm.GLOBAL.appearance ~= appearance then
    wezterm.GLOBAL.appearance = appearance
    wezterm.background_child_process {
        os.getenv "SHELL", -- /usr/local/bin/fish
        "-c",
        "set --universal OS_APPEARANCE " .. appearance,
    }
end

return {}
```
